active_help.go

```diff
@@ -0,0 +1,48 @@
+package cobra
+
+import (
+	"fmt"
+	"os"
+	"strings"
+)
+
+const (
+	activeHelpMarker = "_activeHelp_ "
+	// The below values should not be changed: programs will be using them explicitly
+	// in their user documentation, and users will be using them explicitly.
+	activeHelpEnvVarSuffix  = "_ACTIVE_HELP"
+	activeHelpGlobalEnvVar  = "COBRA_ACTIVE_HELP"
+	activeHelpGlobalDisable = "0"
+)
+
+// AppendActiveHelp adds the specified string to the specified array to be used as ActiveHelp.
+// Such strings will be processed by the completion script and will be shown as ActiveHelp
+// to the user.
+// The array parameter should be the array that will contain the completions.
+// This function can be called multiple times before and/or after completions are added to
+// the array.  Each time this function is called with the same array, the new
+// ActiveHelp line will be shown below the previous ones when completion is triggered.
+func AppendActiveHelp(compArray []string, activeHelpStr string) []string {
+	return append(compArray, fmt.Sprintf("%s%s", activeHelpMarker, activeHelpStr))
+}
```

Comment:
I am not sure with the API here, why do I give the compArray as parameter and get the same value with the appended entry back.

IMO It would make more sense to have it like this:
```go
func AppendActiveHelp(activeHelpStr string) string {
	return fmt.Sprintf("%s%s", activeHelpMarker, activeHelpStr)
}
``` 
and then in you code call
```
comps = append(comps, cobra.AppendActiveHelp("text"))
```

With this the caller has control over the append. While I do not think we need to worry about performance here I believe it is better when you call append only once.
```
text1 := cobra.AppendActiveHelp("text1")
text2 := cobra.AppendActiveHelp("text2")
comps = append(comps, text1, text2)
```

Reply:
I struggled to define a good API to add ActiveHelp messages.  Here is how I remember choosing the `AppendActiveHelp(compArray []string, activeHelpStr string)` API.

1. I was worried that it was not obvious that activeHelp messages should be included in the completions array returned by `ValidArgsFunction`; the two concepts could be seen as independent by users (completions and activeHelp messages).  I therefore felt that by requiring the user to specify an array (and saying in the godoc that the array should be the one also containing the completion choices) we would make it much clearer that activeHelp messages should be included in the array of completions.
2. based on that, I then thought we should mimic the behaviour of Go's `append()` which is why I called it `AppendActiveHelp()` and required the user to assign the result to an array (just like `append()`).  All this in the hope to build on existing knowledge of an existing API.

If end up preferring
```
func AppendActiveHelp(activeHelpStr string) string {
	return fmt.Sprintf("%s%s", activeHelpMarker, activeHelpStr)
}
```
we should probably rename the method to avoid using the word "append".  We could create it as `CreateActiveHelpMsg()` for example.  

But then I worry again that a user won't know easily what to do after having created an activeHelp message.  Will they realize they must include it in the completions array?

Some people had trouble understanding what the arguments to `ValidArgsFunction(cmd, args, toComplete)` represented... So I don't want to assume anything is clear 😄

Reply:
Hmmmmm this is a tricky one. I'm not sure on this.

I think I'm a fan of using the API presented here by Marc, but I'm warry of introducing more complexity to how users get their completions into their program.

But from my understanding of this, it's completely opt in? It's **_not_** another required step users must take to get their completions working?

Reply:
Correct, the existing API for normal completions is not affected at all.  ActiveHelp is 100% opt-in and can be completely ignored by programs using Cobra.